sdk/core/azure-core-amqp/src/main/java/com/azure/core/amqp/implementation/ReactorReceiver.java

```diff
@@ -15,77 +17,145 @@
             .map(state -> {
                 logger.verbose("connectionId[{}] entityPath[{}] linkName[{}] State {}", handler.getConnectionId(),
                     entityPath, getLinkName(), state);
                 return AmqpEndpointStateUtil.getConnectionState(state);
             })
+            .doOnError(error -> {
+                final String message = isDisposed.getAndSet(true)
+                    ? "This was already disposed. Dropping error."
+                    : "Freeing resources due to error.";
+                logger.warning("connectionId[{}] entityPath[{}] linkName[{}] {}",
+                    handler.getConnectionId(), entityPath, getLinkName(), message, error);
+
+                completeClose();
+            })
+            .doOnComplete(() -> {
+                final String message = isDisposed.getAndSet(true)
+                    ? "This was already disposed."
+                    : "Freeing resources.";
+                logger.verbose("connectionId[{}] entityPath[{}] linkName[{}] {}", handler.getConnectionId(),
+                    entityPath, getLinkName(), message);
+
+                completeClose();
+            })
             .cache(1);
 
-        this.subscriptions = this.tokenManager.getAuthorizationResults().subscribe(
-            response -> {
-                logger.verbose("connectionId[{}] linkName[{}] Token refreshed. {}", handler.getConnectionId(),
-                    getLinkName(), response);
-            }, error -> {
-                logger.info("connectionId[{}] entityPath{}] linkName[{}] tokenRenewalFailure[{}]",
-                    handler.getConnectionId(), this.entityPath, getLinkName(), error.getMessage());
-                dispose(new ErrorCondition(Symbol.getSymbol(AmqpErrorCondition.NOT_ALLOWED.getErrorCondition()),
-                    error.getMessage()));
-            }, () -> {
-                logger.verbose("connectionId[{}] entityPath[{}] linkName[{}] Authorization completed. Disposing.",
-                    handler.getConnectionId(), entityPath, getLinkName());
-                dispose();
-            });
+        //@formatter:off
+        this.subscriptions = Disposables.composite(
+            this.endpointStates.subscribe(),
+            this.tokenManager.getAuthorizationResults()
+                .onErrorResume(error -> {
+                    // When we encounter an error refreshing authorization results, close the receive link.
+                    final Mono<Void> operation =
+                        closeAsync(String.format(
+                            "connectionId[%s] linkName[%s] Token renewal failure. Disposing receive link.",
+                            amqpConnection.getId(), getLinkName()),
+                            new ErrorCondition(Symbol.getSymbol(AmqpErrorCondition.NOT_ALLOWED.getErrorCondition()),
+                                error.getMessage()));
+
+                    return operation.then(Mono.empty());
```

Comment:
Can this be done in the `subscribe`'s error handler instead of using `onErrorResume`? It is odd to have an empty error handler below.

Reply:
I had that before but on an error, we want to output another async operation (ie. close the receiver) and the handlers only support synchronous callbacks, I moved it back to be a part of the chain.